presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.log.Logger;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.optimizations.PlanNodeSearcher;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toMap;
+
+public class LocalDynamicFilter
+{
+    private static class Channel
+    {
+        private final int buildChannelIndex;
+        private final DynamicFilters.Descriptor descriptor;
+
+        private Channel(int buildChannelIndex, DynamicFilters.Descriptor descriptor)
+        {
+            this.buildChannelIndex = buildChannelIndex;
+            this.descriptor = descriptor;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("buildChannelIndex", buildChannelIndex)
+                    .add("descriptor", descriptor)
+                    .toString();
+        }
+    }
+
+    private static final Logger log = Logger.get(LocalDynamicFilter.class);
+
+    // Mapping from dynamic filter ID to its probe symbols.
+    private final Multimap<String, Symbol> probeSymbols;
```

Comment:
IIUC, in case there is a union on the probe side, we may need to match several probe-side symbols for a single dynamic filter.
For example, consider:
```
SELECT * 
FROM ((SELECT x FROM f1) UNION (SELECT x FROM f2)) f, d 
WHERE f.x = d.x AND d.y = 1; 
```
In this case, we should use the dynamic filter both for filtering `f1` and `f2`.

I've updated the comments to (hopefully) better explain this scenario. WDYT?